### PR TITLE
[Flyte-7005] Supporting Ray autoscalerOptions

### PR DESCRIPTION
**Pull request by 0yukali0**
## Tracking issue
# Related #7005 

## Why are the changes needed?
Ray autoScalerOptions is not supportted in flyte v1

## What changes were proposed in this pull request?
1. Add autoscalerOptions to ray.proto
2. Generate go files of grpc proto
3. Parsing the options from grpc and translate it to crd of ray operator. 

### Labels

Please add one or more of the following labels to categorize your PR:
- **added**: For new features.
- **changed**: For changes in existing functionality.
- **deprecated**: For soon-to-be-removed features.
- **removed**: For features being removed.
- **fixed**: For any bug fixed.
- **security**: In case of vulnerabilities

This is important to improve the readability of release notes.

### Setup process

1. clone flyte and flytekit (branch flyte-7005 and flytekit-7005)
2. run the flyte backend in sandbox
2.1 flytectl demo start --dev
2.2 make compile && POD_NAMESPACE=flyte ./flyte start --config flyte-single-binary-local.yaml
3. Update ray example like [Ray example](https://github.com/0yukali0/flytesnacks/blob/flytesnack-7005/examples/ray_plugin/ray_plugin/ray_example.py#L68-L74)
3.1 uv venv -p 3.11 && source .venv/bin/activate && uv pip install -e ~/flytekit ~/flytekit/plugins/flytekit-ray && uv pip install -e ~/flyte/flyteidl
3.2 "pyflyte run --remote ray_example.py ray_workflow --n 10"

### Screenshots

image

image

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: flyteidl/protos/flyteidl/plugins/ray.proto [hunk 1]
```diff
@@ -22,6 +22,64 @@ message RayJob {
   string runtime_env_yaml = 5;
 }
 
+message Resources {
```
Comment (machichima) [inline flyteidl/protos/flyteidl/plugins/ray.proto]: Can we use `core.Resources`?

https://github.com/flyteorg/flyte/blob/e5905e7c0ec05ba2700af63160e92c6c92bcac97/flyteidl/protos/flyteidl/core/tasks.proto#L17-L17

File: flyteidl/protos/flyteidl/plugins/ray.proto [hunk 2]
```diff
@@ -22,6 +22,64 @@ message RayJob {
...
+  // Name for config map or secret, container name for resource, path for field
+  string name = 2;
+  // Key for config map or secret, resource name for resource
+  string key = 3;
+}
+
+message EnvVar {
+  string name = 1;
+  string value = 2;
+  EnvValueFrom valueFrom = 3;
```
Comment (machichima) [inline flyteidl/protos/flyteidl/plugins/ray.proto]: Why do we need `EnvValueFrom` here? Can we use `core.KeyValuePair` as the type for env?

https://github.com/flyteorg/flyte/blob/e5905e7c0ec05ba2700af63160e92c6c92bcac97/flyteidl/protos/flyteidl/core/literals.proto#L192-L198

File: flyteidl/protos/flyteidl/plugins/ray.proto [hunk 3]
```diff
@@ -22,6 +22,64 @@ message RayJob {
...
+}
+
+message AutoscalerOptions {
+  // "Default", "Aggressive", "Conservative"
+  string upscaling_mode = 1;
```
Comment (machichima) [inline flyteidl/protos/flyteidl/plugins/ray.proto]: Could we make this as enum?

```proto
enum UpscalingMode {
    UPSCALING_MODE_UNSPECIFIED = 0;
    UPSCALING_MODE_DEFAULT = 1;
    UPSCALING_MODE_AGGRESSIVE = 2;
    UPSCALING_MODE_CONSERVATIVE = 3;
  }
```

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go [hunk 4]
```diff
@@ -129,6 +130,103 @@ func (rayJobResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsC
 	return rayjob, err
 }
 
+func NewAutoscalerOptions(options *plugins.AutoscalerOptions) *rayv1.AutoscalerOptions {
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray.go]: Let's name it as `buildAutoscalerOptions` to follow the naming convention (e.g. like `buildHeadPodTemplate`), and make it private

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go [hunk 5]
```diff
@@ -155,6 +258,7 @@ func constructRayJob(taskCtx pluginsCore.TaskExecutionContext, rayJob *plugins.R
 		},
 		WorkerGroupSpecs:        []rayv1.WorkerGroupSpec{},
 		EnableInTreeAutoscaling: &rayJob.RayCluster.EnableAutoscaling,
+		AutoscalerOptions:       autoScalarOptions,
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray.go]: While protobuf generated `GetXXX()` has `nil` check, we can directly pass:

```go
AutoscalerOptions: NewAutoscalerOptions(rayJob.GetRayCluster().GetAutoscalerOptions())
```

**Comment by machichima** [comment]
Hi @0yukali0,
Could you briefly describe how you test and provide the test script? 
Thank you!

**Comment by 0yukali0** [comment]
Hi @machichima , i updated the setup steps and screenshot in the PR descriptions.
i am going to add the resources field in headspec and workgroupspec to replace requests and limits in next commit.

**Comment by 0yukali0** [comment]
with kuberay 1.1.0 and 1.5.1，run ray task with 2.54.1 is worked.
image
image

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go [hunk 6]
```diff
@@ -129,6 +130,72 @@ func (rayJobResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsC
...
+			autoScalarOptions.Image = &image
+		}
+		if res := options.GetResources(); res != nil {
+			resourceRequirements := v1.ResourceRequirements{}
+			if requests := res.GetRequests(); len(requests) > 0 {
+				resourceRequirements.Requests = convertResourceEntriesToResourceList(requests)
+			}
+			if limits := res.GetLimits(); len(limits) > 0 {
+				resourceRequirements.Limits = convertResourceEntriesToResourceList(limits)
+			}
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray.go:161]: Could we use `flytek8s.ToK8sResourceRequirements` here instead?

https://github.com/flyteorg/flyte/blob/6e84ac2fc3257182e20b6ddca528070ca61f776e/flyteplugins/go/tasks/pluginmachinery/flytek8s/utils.go#L52-L68

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go [hunk 7]
```diff
@@ -129,6 +130,72 @@ func (rayJobResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsC
...
+			autoScalarOptions.Env = []v1.EnvVar{}
+			for _, env := range envs {
+				name := env.GetKey()
+				if val := env.GetValue(); val != "" {
+					autoScalarOptions.Env = append(autoScalarOptions.Env, v1.EnvVar{
+						Name:  name,
+						Value: val,
+					})
+				}
+			}
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray.go:174]: Could we use `flytek8s.ToK8sEnvVar` instead?

https://github.com/flyteorg/flyte/blob/6e84ac2fc3257182e20b6ddca528070ca61f776e/flyteplugins/go/tasks/pluginmachinery/flytek8s/utils.go#L12-L12

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go [hunk 8]
```diff
@@ -129,6 +130,72 @@ func (rayJobResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsC
 	return rayjob, err
 }
 
+func buildAutoscalerOptions(options *plugins.AutoscalerOptions) *rayv1.AutoscalerOptions {
+	var autoScalarOptions *rayv1.AutoscalerOptions
+	if options != nil {
+		autoScalarOptions = &rayv1.AutoscalerOptions{}
+		idleTimeoutTime := options.GetIdleTimeoutSeconds()
+		autoScalarOptions.IdleTimeoutSeconds = &idleTimeoutTime
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray.go:138]: If the `idleTimeoutTime` is not set (= 0), this part will overwrite the default value (60 sec) set in kuberay. we should only set `autoScalarOptions.IdleTimeoutSeconds` if it's not 0 (not empty)

https://github.com/ray-project/kuberay/blob/0706043ae00a0dedfadead423926136e01daaef9/proto/cluster.proto#L145-L147

https://github.com/ray-project/ray/blob/db711158e25231b77ef2fb968320b2319270732d/python/ray/autoscaler/_private/kuberay/autoscaling_config.py#L108-L111

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go [hunk 9]
```diff
@@ -129,6 +130,72 @@ func (rayJobResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsC
 	return rayjob, err
 }
 
+func buildAutoscalerOptions(options *plugins.AutoscalerOptions) *rayv1.AutoscalerOptions {
+	var autoScalarOptions *rayv1.AutoscalerOptions
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray.go:134]:
```suggestion
	var autoScalerOptions *rayv1.AutoscalerOptions
```

nit

File: flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go [hunk 10]
```diff
@@ -1592,3 +1592,129 @@ func transformStructToStructPB(t *testing.T, obj interface{}) *structpb.Struct {
...
+			assert.Len(t, result, tt.wantLen)
+			for _, key := range tt.wantKeys {
+				_, ok := result[key]
+				assert.True(t, ok, "expected key %s in result", key)
+			}
+		})
+	}
+}
+
+func TestNewAutoscalerOptions(t *testing.T) {
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go:1652]:
```suggestion
func TestBuildAutoscalerOptions(t *testing.T) {
```
nit

File: flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go [hunk 11]
```diff
@@ -1592,3 +1592,129 @@ func transformStructToStructPB(t *testing.T, obj interface{}) *structpb.Struct {
...
+		})
+		require.NotNil(t, result)
+		require.NotNil(t, result.Resources)
+		assert.Equal(t, resource.MustParse("250m"), result.Resources.Requests[corev1.ResourceCPU])
+		assert.Equal(t, resource.MustParse("512Mi"), result.Resources.Requests[corev1.ResourceMemory])
+		assert.Equal(t, resource.MustParse("1"), result.Resources.Limits[corev1.ResourceCPU])
+		assert.Equal(t, resource.MustParse("1Gi"), result.Resources.Limits[corev1.ResourceMemory])
+	})
+
+	t.Run("env literal value", func(t *testing.T) {
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go:1711]: Could we add a test for env with empty value (should be skipped)?

File: flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go [hunk 12]
```diff
@@ -1592,3 +1592,129 @@ func transformStructToStructPB(t *testing.T, obj interface{}) *structpb.Struct {
...
+		})
+	}
+}
+
+func TestNewAutoscalerOptions(t *testing.T) {
+	t.Run("nil input returns nil", func(t *testing.T) {
+		assert.Nil(t, buildAutoscalerOptions(nil))
+	})
+
+	t.Run("idle timeout propagated", func(t *testing.T) {
```
Comment (machichima) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go:1657]: Could we also add a test like this:

```go
t.Run("idle timeout zero should not be set", func(t *testing.T) {
      result := buildAutoscalerOptions(&plugins.AutoscalerOptions{})
      require.NotNil(t, result)
      assert.Nil(t, result.IdleTimeoutSeconds)
  })
```

Which is related to my comment above: https://github.com/flyteorg/flyte/pull/7111/changes#r3114611720
If the `IdleTimeoutSeconds` is not set, it should remain empty so that Ray can apply its default value

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go (hunk 8 above)
Comment (0yukali0) [reply flyteplugins/go/tasks/plugins/k8s/ray/ray.go:138]: Hi @machichima
I will update it and the following test case to check the setting.

File: flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go (hunk 11 above)
Comment (0yukali0) [reply flyteplugins/go/tasks/plugins/k8s/ray/ray_test.go:1705]: flytek8s.ToK8sEnvVar allows empty value, so i will skip the test with empty value.

File: flyteplugins/go/tasks/plugins/k8s/ray/ray.go [hunk 13]
```diff
@@ -129,6 +130,55 @@ func (rayJobResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsC
...
+	return autoScalerOptions
+}
+
+func convertResourceEntriesToResourceList(entries []*core.Resources_ResourceEntry) v1.ResourceList {
+	resourceList := v1.ResourceList{}
+	for _, entry := range entries {
+		var name v1.ResourceName
+		switch entry.GetName() {
+		case core.Resources_CPU:
+			name = v1.ResourceCPU
```
Comment (pingsutw) [inline flyteplugins/go/tasks/plugins/k8s/ray/ray.go]: Do we use it anywhere? if not, can we remove it

File: flyteidl/protos/flyteidl/plugins/ray.proto [hunk 14]
```diff
@@ -22,6 +23,23 @@ message RayJob {
...
+    UPSCALING_MODE_CONSERVATIVE = 3;
+  }
+  UpscalingMode upscaling_mode = 1;
+  int32 idle_timeout_seconds = 2;
+  // autoscaler sidecar env vars
+  repeated core.KeyValuePair env = 3;
+  // custom autoscaler image
+  string image = 4;
+  // autoscaler container resources
+  core.Resources resources = 5;
```
Comment (pingsutw) [inline flyteidl/protos/flyteidl/plugins/ray.proto:40]: Should we use `core.K8sPod k8s_pod` instead of `core.Resources`? SDk can use this podSpec to override to resource, label or annotation. same as HeadGroupSpec and WorkerGroupSpec

Comment (0yukali0) [reply flyteidl/protos/flyteidl/plugins/ray.proto:40]: No, AutoscalerOptions can't follow same solution as  HeadGroupSpec and WorkerGroupSpec, the reason is following.
The [api doc](https://github.com/ray-project/kuberay/blob/master/ray-operator/apis/ray/v1alpha1/raycluster_types.go) describes that [AutoscalerOptions ](https://github.com/ray-project/kuberay/blob/master/ray-operator/apis/ray/v1alpha1/raycluster_types.go#L75) currently only support corev1.ResourceRequirements  when [HeadGroupSpec ](https://github.com/ray-project/kuberay/blob/master/ray-operator/apis/ray/v1alpha1/raycluster_types.go#L33)and [WorkerGroupSpec](https://github.com/ray-project/kuberay/blob/master/ray-operator/apis/ray/v1alpha1/raycluster_types.go#L60) support corev1.PodTemplateSpec.

Comment (0yukali0) [reply flyteidl/protos/flyteidl/plugins/ray.proto:40]: Same situation also exists in [v1 version](https://github.com/ray-project/kuberay/blob/master/ray-operator/apis/ray/v1/raycluster_types.go#L210-L253)

**Review by pingsutw — APPROVED** [review]
